File: lib/src/main/java/eu/bakici/imageprogressbar/indicator/CircularIndicator.java
package eu.bakici.imageprogressbar.indicator;

/*
 * Copyright (C) 2016 hayribakici
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;
import android.graphics.Shader;

import androidx.annotation.FloatRange;
import androidx.annotation.IntDef;
import androidx.annotation.NonNull;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import eu.bakici.imageprogressbar.utils.IndicatorUtils;

public class CircularIndicator extends ProgressIndicator {


    public static final int CLOCKWISE = 0;
    public static final int COUNTERCLOCKWISE = 1;
    protected static final int FULL_CIRCLE = 360;

    /**
     * Type of how the image will be processed.
     */
    @Retention(RetentionPolicy.SOURCE)
    @IntDef(value = {
            CLOCKWISE,
            COUNTERCLOCKWISE
    })
    public @interface Turn {

    }

    private final Paint paint;
    private BitmapShader coloredShader;
    private BitmapShader bwShader;
    private RectF arc;

    private final int turn;

    public CircularIndicator() {
        this(CLOCKWISE);
    }

    public CircularIndicator(@Turn int turn) {
        super();
        this.turn = turn;
        paint = new Paint();
    }


    @Override
    public Bitmap getPreProgressBitmap(@NonNull Bitmap originalBitmap) {
        coloredShader = new BitmapShader(originalBitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP);
        arc = new RectF(originalBitmap.getWidth() * -0.5f,
                originalBitmap.getHeight() * -0.5f,
                originalBitmap.getWidth() * 1.5f,
                originalBitmap.getHeight() * 1.5f);
        Bitmap bwBitmap = IndicatorUtils.convertGrayscale(originalBitmap);
        bwShader = new BitmapShader(bwBitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP);

        return bwBitmap;
    }


    @Override
    public Bitmap getBitmap(@NonNull Bitmap originalBitmap, @FloatRange(from = 0.0, to = 1.0) float progressPercent) {
        int angle = IndicatorUtils.getValueOfPercent(FULL_CIRCLE, progressPercent);
        if (turn == COUNTERCLOCKWISE) {
            angle = angle * (-1);
        }
        Bitmap bitmap = Bitmap.createBitmap(originalBitmap.getWidth(), originalBitmap.getHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        // drawing the colored arc with its counter arc
        drawArc(canvas, 270, (FULL_CIRCLE - angle) * (-1), bwShader);
        drawArc(canvas, 270, angle, coloredShader);
        return bitmap;
    }

    private void drawArc(@NonNull Canvas canvas, int startAngle, int angle, BitmapShader shader) {
        Paint paint = new Paint();
        paint.setShader(shader);
        canvas.drawArc(arc, startAngle, angle, true, paint);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.DST_ATOP));
    }
}

File: lib/src/main/java/eu/bakici/imageprogressbar/indicator/CircularIndicator.kt
/*
 * Copyright (C) 2016, 2022 hayribakici
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package eu.bakici.imageprogressbar.indicator

import android.graphics.*
import androidx.annotation.IntDef
import androidx.annotation.NonNull
import eu.bakici.imageprogressbar.utils.IndicatorUtils


class CircularIndicator(@Turn private val turn: Int = CLOCKWISE) : Indicator() {

    companion object {
        const val CLOCKWISE = 0
        const val COUNTERCLOCKWISE = 1
        private const val FULL_CIRCLE = 360
    }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(value = [CLOCKWISE, COUNTERCLOCKWISE])
    annotation class Turn

    private var coloredShader: BitmapShader? = null
    private var bwShader: BitmapShader? = null
    private var arc: RectF? = null

    override fun getPreProgressBitmap(originalBitmap: Bitmap): Bitmap {
        coloredShader = BitmapShader(originalBitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
        // stretching the canvas of the arc
        arc = RectF(originalBitmap.width * -0.5f,
                originalBitmap.height * -0.5f,
                originalBitmap.width * 1.5f,
                originalBitmap.height * 1.5f)
        val bwBitmap = IndicatorUtils.convertGrayscale(originalBitmap)
        bwShader = BitmapShader(bwBitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
        return bwBitmap
    }

    override fun getBitmap(state: ProgressState): Bitmap {
        val progress = state.progress
        var angle = IndicatorUtils.getValueOfPercent(FULL_CIRCLE, progress)
        if (turn == COUNTERCLOCKWISE) {
            angle = -angle
        }
        val originalBitmap = state.originalBitmap!!
        val bitmap = Bitmap.createBitmap(originalBitmap.width, originalBitmap.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        // drawing the colored arc with its counter arc
        drawArc(canvas, -(FULL_CIRCLE - angle), bwShader!!)
        drawArc(canvas, angle, coloredShader!!)
        return bitmap
    }

    private fun drawArc(@NonNull canvas: Canvas, angle: Int, @NonNull shader: BitmapShader) {
        val paint = Paint()
        paint.shader = shader
        canvas.drawArc(arc!!, 270f, angle.toFloat(), true, paint)
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.DST_ATOP)
    }

}